TrkPatRec/src/RobustMultiHelixFinder_module.cc

```diff
@@ -0,0 +1,1040 @@
+    minDXY2Circle_   (config().minDXY2Circle()),
+    maxDXY2Circle_   (config().maxDXY2Circle()),
+    targetCon_       (config().targetCon()),
+    targetRadius_    (config().targetRadius()),
+    ccMinEnergy_     (config().ccMinEnergy()),
+    ccWeight_        (config().ccWeight()),
+    minDPDZSlope_    (config().minDPDZSlope()),
+    maxDPDZSlope_    (config().maxDPDZSlope()),
+    DPDZStep_        (config().DPDZStep()),
+    MaxDPhiHelInit_  (config().maxDPhiHelInit()),
+    maxDPhiHelFit_   (config().maxDPhiHelFit()),
+    maxDTHelFit_     (config().maxDTHelFit()),
+    maxChi2Hit_      (config().maxChi2Hit()),
+    minDZTrk_        (config().minDZTrk()),
+    fitCircleAlg_    (config().fitCircleAlg()),
+    minStrawHits_    (std::max(config().minStrawHits(),3u)),
+    nMaxTrkIter_     (config().nMaxTrkIter()),
+    iev_             (0),
+    diag_            (config().diagLevel()),
+    diagTool_(),
+    data_()
+  {
+     std::vector<int> helvals = config().helicities();
+     for (auto hv : helvals) {
+       Helicity hel(hv);
+       hels_.emplace_back(hel);
+       produces<HelixSeedCollection>(Helicity::name(hel));
+     }
+
+     if (diag_) diagTool_ = art::make_tool<ModuleHistToolBase>(config().diagPlugin," ");
+  }
+
+
+  //--------------------------------------------------------------------------------------------------------------
+  void RobustMultiHelixFinder::beginJob()
+  {
+    if (diag_){
+       art::ServiceHandle<art::TFileService> tfs;
+       diagTool_->bookHistograms(tfs);
+    }
+  }
+
+
+  //---------------------------------------------------------------------------------------------------------------------------
+  void RobustMultiHelixFinder::produce(art::Event& event )
+  {
+     if (diag_>0) std::cout<<"Event "<<event.id().event()<<std::endl;
+     iev_ = event.id().event();
+
+     const auto& tcH = event.getValidHandle(tcToken_);
+     const auto& chH = event.getValidHandle(chToken_);
+
+     mapHelix helcols;
+     for (const auto& hel : hels_) helcols[hel] = std::unique_ptr<HelixSeedCollection>(new HelixSeedCollection());
+
+     findAllHelices(event, helcols,tcH, chH);
+
+     for (const auto& hel : hels_) event.put(std::move(helcols[hel]),Helicity::name(hel));
+  }
+
+
+
+  //---------------------------------------------------------------------------------------------------------------------------
+  // Find all helices in an event (hopefully), some timeCluster implementations have overlapping content, so filter duplicates
+  void RobustMultiHelixFinder::findAllHelices(art::Event& event, mapHelix& helcols, const art::ValidHandle<TimeClusterCollection>& tcH,
+                                              const art::ValidHandle<ComboHitCollection>& chH)
+  {
+    cal_ = &(*GeomHandle<Calorimeter>());
+
+    const TimeClusterCollection& tccol(*tcH);
+    const ComboHitCollection&    chcol(*chH);
+
+    if (diag_) {data_.reset(); data_.event_=&event; data_.chcol_ = &chcol;}
```

Comment:
I am not a fan of using tools for MC diagnostics for 2 reasons: 1) it forces you to do tricks like this cache of the event inside of a data struct and 2) you can only run diagnostics in the producer, not on output collections read from data files.  Sometimes it's necessary, as when you make diagnostics on intermediate products (not just the final products), but here it seems you are running diagnostics on the same products as will be put into the event.  Did you considered making this tool into an analyzer module instead?

Reply:
I was under the impression that this was the standard solution, but I might have been wrong. I need to check if the diagnosis uses information that isn't captured in the data product, but if this isn't the case the an analyzer module would be a possibility.